examples/pytorch_lightning/swav_queue.py

```diff
@@ -0,0 +1,120 @@
+# from the paper. The settings are chosen such that the example can easily be
+# run on a small dataset with a single GPU.
+
+import torch
+from torch import nn
+import torchvision
+import pytorch_lightning as pl
+
+from lightly.data import LightlyDataset
+from lightly.data import SwaVCollateFunction
+from lightly.loss import SwaVLoss
+from lightly.loss.memory_bank import MemoryBankModule
+from lightly.models.modules import SwaVProjectionHead
+from lightly.models.modules import SwaVPrototypes
+
+
+class SwaV(pl.LightningModule):
+    def __init__(self):
+        super().__init__()
+        resnet = torchvision.models.resnet18()
+        self.backbone = nn.Sequential(*list(resnet.children())[:-1])
+        self.projection_head = SwaVProjectionHead(512, 512, 128)
+        self.prototypes = SwaVPrototypes(
+            128, 512, 10
+        )
+        self.start_queue_at_epoch = 15
+        self.queues = nn.ModuleList([MemoryBankModule(size=3840) for _ in range(2)])
+        self.criterion = SwaVLoss()
+        self.step = 0
+
+    def _subforward(self, input):
+        features = self.backbone(input).flatten(start_dim=1)
+        features = self.projection_head(features)
+        features = nn.functional.normalize(features, dim=1, p=2)
+        return features
+
+    @torch.no_grad()
+    def _get_queue_prototypes(self, high_resolution_features):
+        if self.queues is None:
+            return None
+
+        if len(high_resolution_features) != len(self.queues):
+            raise ValueError(
+                f"The number of queues ({len(self.queues)}) should be equal to the number of high "
+                f"resolution inputs ({len(high_resolution_features)}). Set `n_queues` accordingly."
+            )
+
+        # Get the queue features
+        queue_features = []
+        for i in range(len(self.queues)):
+            _, features = self.queues[i](high_resolution_features[i], update=True)
+            # Queue features are in (num_ftrs X queue_length) shape, while the high res
+            # features are in (batch_size X num_ftrs). Swap the axes for interoperability.
+            features = torch.permute(features, (1, 0))
+            queue_features.append(features)
+
+        # If loss calculation with queue prototypes starts at a later epoch,
+        # just queue the features and return None instead of queue prototypes.
+        if self.start_queue_at_epoch > 0:
+            if self.current_epoch is None:
+                raise ValueError(
+                    "The epoch number must be passed to the `forward()` "
+                    "method if `start_queue_at_epoch` is greater than 0."
+                )
+            if self.current_epoch < self.start_queue_at_epoch:
+                return None
+
+        # Assign prototypes
+        queue_prototypes = [self.prototypes(x) for x in queue_features]
+        return queue_prototypes
+
+    def training_step(self, batch, batch_idx):
```

Comment:
Let's move the `training_step` and `configure_optimizers` methods just below the `__init__` method. It is best to put first the public methods and then the "private" methods starting with an underscore.